### ISSUE TITLE
How can I train Image Classification on custom cnn layers ?

### ISSUE THREAD

**Issue by rhlr** [ludwig-ai/ludwig#499]
I have a model with my own custom layers. How can I use `ludwig` to train image classification using those layers ?

**Comment by carlogrisetti** [comment]
You can recreate your model definition in the yaml passed to ludwig, as long as it's based on a convolutional stack encoder or a resnet template. There are default values but you can customize the shape of the network by specifying layers \ filter size \ etc.
Refer to https://uber.github.io/ludwig/user_guide/#image-input-features-and-encoders
If you have your own model "structure" you'd like to implement, that could be a solution depending on how detailed you need to be.

**Comment by rhlr** [comment]
So, let's suppose for example, this is my model:

    model = Sequential()
    model.add(Conv2D(32, kernel_size=(3, 3),
                activation='relu',
                input_shape=(28,28,1)))
    model.add(Conv2D(64, (3, 3), activation='relu'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Dropout(0.25))
    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(num_classes, activation='softmax'))

So how can I create the model definition file for this model ? I cannot see all the options in stacked or resnet template.

**Comment by carlogrisetti** [comment]
One thing you cannot do, for example, is to set different dropout rates for different layers. You can only set a "global" rate, and then decide to which feature you want to apply that
convolutional size, number of filters, kernel size, activation functions and everything else can be more or less be done.
There won't be all options, that's why I was saying that this could be done depending on how detailed you want\need to be in matching your own model definition.

I suppose you could extend ludwig by implementing your own custom feature that builds the model exactly as you want it, but I don't know if this is your intention. In that case: https://uber.github.io/ludwig/developer_guide/

**Comment by msaisumanth** [comment]
@rhlr the only issue here would be the different dropout rates as @carlogrisetti mentioned. 

The rest of it can be specified in the yaml. Look at this example: https://uber.github.io/ludwig/examples/#image-classification-mnist

it uses 2 convolutional layers.  You can also specify height, width etc. like below

```yaml
input_features:
    -
        name: image_path
        type: image
        encoder: stacked_cnn
        preprocessing: 
               width: 28
               height: 28


```

**Comment by w4nderlust** [comment]
The different dropout rate is coming in the future.
Moreover, if you need to implement your own encoder in a really custom way, you can follow the instructions in the developers guide to add your own. It's really easy, as long as you get a tensor in and provide a tensor out, you can define whatever model you want, also using tf.keras layers.
